membership_product_set/models/res_partner.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2019 Yu Weng <yweng@elegosoft.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, models, _
+from odoo.exceptions import UserError
+
+
+class ResPartner(models.Model):
+    _inherit = "res.partner"
+
+    @api.multi
+    def create_membership_invoice(self, product_id=None, datas=None):
+        product_id = product_id or datas.get('membership_product_id')
+        amount = datas.get('amount', 0.0)
+        product = self.env['product.product'].browse(product_id)
+        invoice_list = []
+        if product.membership_set:
+            for partner in self:
+                if partner.free_member:
+                    raise UserError(_("Partner is a free Member."))
+                account_id = partner.property_account_receivable_id.id
+                w = _("Partner doesn't have an account to make the invoice.")
+                if not account_id:
+                    raise UserError(w)
+                position_id = partner.property_account_position_id.id
+                invoice = self.env['account.invoice'].create({
+                    'partner_id': partner.id,
+                    'account_id': account_id,
+                    'fiscal_position_id': position_id
+                })
+                for product in product.membership_set_products:
+                    price_dict = product.price_compute('list_price')
+                    amount = price_dict.get(product.id) or 0
+                    line_values = {
+                        'product_id': product.id,
+                        'price_unit': amount,
+                        'invoice_id': invoice.id,
+                    }
+
+                    # create a record in cache, apply onchange
+                    # then revert back to a dictionnary
+                    invoice_line = self.env['account.invoice.line']\
+                        .new(line_values)
+                    invoice_line._onchange_product_id()
```

Comment:
Why do you need all of this after the explicit call of  **_onchange_product_id** ?

```
                    line_values = invoice_line._convert_to_write({
                        name: invoice_line[name]
                        for name in invoice_line._cache
                    })
                   line_values['price_unit'] = amount
```

It is just needed in this context or always after changing the product_id. In that case it would be better to extend the method **_onchange_product_id** here in this module.

Also if the behavior of changing the name and the price unit is already covered by other onchange, then it might be better not to call explicitly the method of **_onchange_product_id** , note that this method is protected and not guaranteed to work always.

The ideally would be to call the generic *onchange* method like [here](https://github.com/odoo/odoo/blob/11.0/odoo/addons/test_new_api/tests/test_onchange.py#L146) so that Odoo can decide which onchange methods are needed and in which order.

Reply:
but this is also a problem of the original method that is being extended here.

Reply:
Because invoice_line is in that moment only in cache. It is good for the performance.